middle_end/flambda2/to_cmm/to_cmm_primitive.ml

```diff
@@ -130,66 +130,71 @@ let block_load ~dbg (kind : P.Block_access_kind.t) (mutability : Mutability.t)
+  | Mixed { field_kind = Flat_suffix field_kind; shape; _ } ->
+    let chunk : Cmm.memory_chunk =
+      match field_kind with
+      | Tagged_immediate -> Word_int
+      | Naked_float -> Double
+      | Naked_float32 -> Single { reg = Float32 }
+      | Naked_int32 -> Thirtytwo_signed
+      | Naked_vec128 -> Onetwentyeight_unaligned
+      | Naked_int64 | Naked_nativeint -> Word_int
+    in
+    get_field_unboxed chunk
+      ~offset_in_words:
+        (Flambda_kind.Mixed_block_shape.offset_in_words shape field)
 
 let block_set ~dbg (kind : P.Block_access_kind.t) (init : P.Init_or_assign.t)
     ~block ~field ~new_value =
   let init_or_assign = P.Init_or_assign.to_lambda init in
   let field = Targetint_31_63.to_int field in
-  let set_func, offset =
-    match kind with
-    | Mixed { field_kind = Value_prefix Any_value; _ }
-    | Values { field_kind = Any_value; _ } ->
-      C.setfield_computed Pointer init_or_assign, field
-    | Mixed { field_kind = Value_prefix Immediate; _ }
-    | Values { field_kind = Immediate; _ } ->
-      C.setfield_computed Immediate init_or_assign, field
-    | Naked_floats _ -> C.float_array_set, field
-    | Mixed { field_kind = Flat_suffix field_kind; shape; _ } ->
-      let func =
-        (* CR-someday mslater: make these functions consistent *)
-        match field_kind with
-        | Tagged_immediate -> C.setfield_computed Immediate init_or_assign
-        | Naked_float -> C.float_array_set
-        | Naked_float32 -> C.setfield_unboxed_float32
-        | Naked_int32 -> C.setfield_unboxed_int32
-        | Naked_vec128 ->
-          fun arr index_in_words newval dbg ->
-            C.setfield_unboxed_vec128 arr ~index_in_words newval dbg
-        | Naked_int64 | Naked_nativeint -> C.setfield_unboxed_int64_or_nativeint
-      in
-      let offset = Flambda_kind.Mixed_block_shape.offset_in_words shape field in
-      func, offset
+  let setfield_computed is_ptr =
+    let index = C.int_const dbg field in
+    C.return_unit dbg
+      (C.setfield_computed is_ptr init_or_assign block index new_value dbg)
   in
-  let index = C.int_const dbg offset in
-  C.return_unit dbg (set_func block index new_value dbg)
+  match kind with
+  | Mixed { field_kind = Value_prefix Any_value; _ }
+  | Values { field_kind = Any_value; _ } ->
+    setfield_computed Pointer
+  | Mixed { field_kind = Value_prefix Immediate; _ }
+  | Values { field_kind = Immediate; _ } ->
+    setfield_computed Immediate
+  | Naked_floats _ ->
+    let index = C.int_const dbg field in
+    C.float_array_set block index new_value dbg
+  | Mixed { field_kind = Flat_suffix field_kind; shape; _ } ->
+    let memory_chunk : Cmm.memory_chunk =
+      match field_kind with
+      | Tagged_immediate -> Word_int
+      | Naked_float -> Double
+      | Naked_float32 -> Single { reg = Float32 }
+      | Naked_int32 -> Thirtytwo_signed
+      | Naked_vec128 -> Onetwentyeight_unaligned
+      | Naked_int64 | Naked_nativeint -> Word_int
```

Comment:
this code could be factored out into a function from the `field_kind` to the `memory_chunk` and used for both the `get` and `set` primitives.